### ISSUE TITLE
Avoid parallel cleanup jobs

### ISSUE THREAD

**Issue by mattelacchiato** [tomorrow-one/transactional-outbox#304]
As stated in the [Readme](https://github.com/tomorrow-one/transactional-outbox?tab=readme-ov-file#how-to-house-keep-your-outbox-table), you can trigger the cleanup job.

In a distributed setup, the using application needs to elect some kind of leader to avoid parallel execution of this cleanup job. But this library has already a mechanism built in. It would be awesome if transactional-outbox could
a) expose the `active` flag so the caller can check if it's on the current active node or
b) handle the cleanup on its own (with configurable deletion interval and instant).

I would prefer option b), since the library owns these tables anyway...

WDYT?

If you like this idea, I could provide a MR for this...

**Comment by magro** [comment]
I'd love to see a PR for option b! 😀

WDYT @hnrkdmsk @danielrehmann?

**Comment by danielrehmann** [comment]
Hey👋,
I think I like the idea having a potential independent way of scheduling the clean up natively in this library as option b describes.

When we introduced the cleanup initially we did not see the need of having the library manage the cleanup, as all of our services already have a scheduling library in place that supports distributed services.

If you have the need of introducing this library without having such scheduling possibilities in place the proposal seems nice and would be open for such an addition in the library.

It would be important to me that it is disabled by default and we still have the possibility of cleaning the tables from an outside scheduling job.

Thanks for the proposal 👍

**Comment by magro** [comment]
@danielrehmann

> It would be important to me that it is disabled by default and we still have the possibility of cleaning the tables from an outside scheduling job.

I assume that you're asking for backwards compatibility, which I can fully understand.
Not sure what you think would be the best default config for new users, starting with the transactional-outbox from scratch?
I think the best default for new users would be automatic cleanup.

Maybe we can
* add a new constructor which takes additional parameters for cleanup interval and duration to keep already processed records - for automatic cleanup
* deprecate the existing constructor, which would not trigger automatic cleanup

WDYT?

**Comment by danielrehmann** [comment]
It's not related to backwards compatibility. I think most systems where the outbox is used most likely already have a distributed scheduling capability with metrics and convenience features in place. Having a cleanup job there is preferable in my opinion as it would be a unified logic of: I have a regular job, it's like any other job.

I would be fine with introducing an additional flag that allows enabling self cleaning but not deprecating or suggesting that this would be a better way of doing it.

**Comment by magro** [comment]
Ok, then we have diverging assumptions. Why do you think most systems where the outbox is used most likely already have a distributed scheduling capability? In my projects of the last years this is not the case: most microservices are really stateless (apart from the database). 

Even if systems would have a distributed scheduling capability in place, additional effort would be required, while a solution where transactional-outbox does cleanup on its own (in a more complete, self-contained manner) would create less effort to introduce transactional-outbox.
In short, I see a difference of required effort, why a prefer option b as the default. After all, what is important to me is that new users don't forget to setup cleanup. 

Because you mention "metrics and convenience features": what do you find important there, could you please elaborate that a bit?

**Comment by danielrehmann** [comment]
As mentioned before I am totally fine with introducing it. But deprecating the non cleanup way is really not the way I would like to see.

Having a default self contained cleaning mechanism is a nice feature. But if there is already a distributed scheduling tool in place, it usually comes already with spring actuator integration where you can see when jobs are scheduled and when they were last executed, as well as metric integration about the job that was run as well as the result.

Every project I worked on in the last 5 years already had such a distributed scheduling in place. I introduced the cleanup of the outbox there with very little effort in all current projects where it is used and I would love to keep it there even when the outbox would have self cleaning capabilities.

**Comment by magro** [comment]
Thanks for pointing to spring actuator integration - do you have a link for that?

**Comment by danielrehmann** [comment]
Quartz would be one option, the other we commonly use is the https://github.com/kagkarlsson/db-scheduler that also comes as a spring boot starter with Health check and Metrics.

**Comment by hnrkdmsk** [comment]
I'd also like to have the self-cleaning job available within this library. However, I'd also not like to have the current way deprecated because this should be an option if you want to have more control about the cleaning process if needed. We should offer both solutions. The self-cleaning job should be able to be disabled. But for me, it can be enabled by default.

**Comment by hnrkdmsk** [comment]
Question: What is the status here, and what are the open questions? 🤗 I see, that we all somehow agree in having the cleanup process implemented in the library. However, we need to find the way how it should be implemented. 

One suggestion is to implement it the same way we did for processing events. Another thought is using third party libraries here. 

Is it possible to use the spring scheduling as an abstraction which should be configured on client side? With this we would not decide in the library what dependency the client has to use here.

**Comment by mattelacchiato** [comment]
> One suggestion is to implement it the same way we did for processing events. Another thought is using third party libraries here.

What are the downsides of "to implement it the same way we did for processing events"? Wouldn't this be the simplest approach?

**Comment by magro** [comment]
I just submitted PR #334 as a proposal to resolve this, looking forward to your feedback there!
